drivers/crypto/crypto_si32.c

```diff
@@ -0,0 +1,1170 @@
+/*
+ * Copyright (c) 2024 GARDENA GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Design decisions:
+ *  - As there is only one AES controller, this implementation is not using a device configuration.
+ *
+ * Notes:
+ *  - If not noted otherwise, chapter numbers refer to the SiM3U1XX/SiM3C1XX reference manual
+ *    (SiM3U1xx-SiM3C1xx-RM.pdf, revision 1.0)
+ *  - Each DMA channel has one word of unused data (=> 3 x 4 = 12 bytes of unused RAM)
+ */
+
+#define DT_DRV_COMPAT silabs_si32_aes
+
+#define LOG_LEVEL CONFIG_CRYPTO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(aes_silabs_si32);
+
+#include <zephyr/crypto/crypto.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <SI32_AES_A_Type.h>
+#include <SI32_CLKCTRL_A_Type.h>
+#include <SI32_DMACTRL_A_Type.h>
+#include "SI32_DMAXBAR_A_Type.h"
+#include <si32_device.h>
+
+#include <errno.h>
+
+#define AES_KEY_SIZE   16
+#define AES_BLOCK_SIZE 16
+
+#define DMA_CHANNEL_COUNT DT_PROP(DT_INST(0, silabs_si32_dma), dma_channels)
+
+#define DMA_CHANNEL_ID_RX  DT_INST_DMAS_CELL_BY_NAME(0, rx, channel)
+#define DMA_CHANNEL_ID_TX  DT_INST_DMAS_CELL_BY_NAME(0, tx, channel)
+#define DMA_CHANNEL_ID_XOR DT_INST_DMAS_CELL_BY_NAME(0, xor, channel)
+
+BUILD_ASSERT(DMA_CHANNEL_ID_RX < DMA_CHANNEL_COUNT, "Too few DMA channels");
+BUILD_ASSERT(DMA_CHANNEL_ID_TX < DMA_CHANNEL_COUNT, "Too few DMA channels");
+BUILD_ASSERT(DMA_CHANNEL_ID_XOR < DMA_CHANNEL_COUNT, "Too few DMA channels");
+
+struct crypto_session {
+	/* Decryption key needed only by ECB and CBC, and counter only by CTR. */
+	union {
+		uint8_t decryption_key[32]; /* only used for decryption sessions */
+		uint32_t current_ctr;       /* only used for AES-CTR sessions */
+	};
+
+	bool in_use;
+};
+
+struct crypto_data {
+	struct crypto_session sessions[CONFIG_CRYPTO_SI32_MAX_SESSION];
+};
+
+K_MUTEX_DEFINE(crypto_si32_in_use);
+K_SEM_DEFINE(crypto_si32_work_done, 0, 1);
+
+static struct crypto_data crypto_si32_data;
+
+static void crypto_si32_dma_completed(const struct device *dev, void *user_data, uint32_t channel,
+				      int status)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(user_data);
+
+	const char *const result = status == DMA_STATUS_COMPLETE ? "succeeded" : "failed";
+
+	switch (channel) {
+	case DMA_CHANNEL_ID_RX:
+		LOG_DBG("AES0 RX DMA channel %s", result);
+		k_sem_give(&crypto_si32_work_done);
+		break;
+	case DMA_CHANNEL_ID_TX:
+		LOG_DBG("AES0 TX DMA channel %s", result);
+		break;
+	case DMA_CHANNEL_ID_XOR:
+		LOG_DBG("AES0 XOR DMA channel %s", result);
+		break;
+	}
```

Comment:
According with the coding guidelines, it needs to have a `default` label [see here](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#id29)

Reply:
Done. Tested by putting `zephyr_compile_options_ifdef(CONFIG_CRYPTO_SI32 -Wswitch-default)` temporarily into `drivers/crypto/CMakeLists.txt`.